components/lora/src/esp_api_impl.c

```diff
@@ -0,0 +1,195 @@
+#include "api/driver_api.h"
+#include "driver/spi_master.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "esp_system.h"
+#include "esp_log.h"
+#include "driver/gpio.h"
+
+static spi_device_handle_t __spi;
+static const char *LORA_API_TAG = "LORA_API";
+
+api_status_t spi_init(void)
+{
+    ESP_LOGI(LORA_API_TAG, "Initializing spi...\n");
+    esp_err_t ret;
+
+    ret = gpio_reset_pin(CONFIG_RST_GPIO);
+    ret += gpio_set_direction(CONFIG_RST_GPIO, GPIO_MODE_OUTPUT);
+
+    lora_reset();
+
+    ret += gpio_reset_pin(CONFIG_CS_GPIO);
+    ret += gpio_set_direction(CONFIG_CS_GPIO, GPIO_MODE_OUTPUT);
+    ret += gpio_set_level(CONFIG_CS_GPIO, 1);
+
+    spi_bus_config_t bus = {
+        .miso_io_num = CONFIG_MISO_GPIO,
+        .mosi_io_num = CONFIG_MOSI_GPIO,
+        .sclk_io_num = CONFIG_SCK_GPIO,
+        .quadwp_io_num = -1,
+        .quadhd_io_num = -1,
+        .max_transfer_sz = 0};
+
+    ret += spi_bus_initialize(SPI2_HOST, &bus, SPI_DMA_CH_AUTO);
+
+    spi_device_interface_config_t dev = {
+        .clock_speed_hz = 9000000,
+        .mode = 0,
+        .spics_io_num = CONFIG_CS_GPIO,
+        .queue_size = 7,
+        .flags = 0,
+        .pre_cb = NULL};
+
+    ret += spi_bus_add_device(SPI2_HOST, &dev, &__spi);
+
+    if (ESP_OK == ret)
+    {
+        ESP_LOGI(LORA_API_TAG, "SPI initialized successfully");
+        return API_OK;
+    }
+    else
+    {
+        ESP_LOGE(LORA_API_TAG, "SPI initialization failed");
+        return API_SPI_ERROR;
+    }
+}
+
+api_status_t spi_write(uint8_t reg, uint8_t val)
+{
+    uint8_t out[2] = {0x80 | reg, val};
+    uint8_t in[2];
+
+    spi_transaction_t t = {
+        .flags = 0,
+        .length = 8 * sizeof(out),
+        .tx_buffer = out,
+        .rx_buffer = in};
+
+    if (ESP_OK == spi_device_transmit(__spi, &t))
+    {
+        return API_OK;
+    }
+    else
+    {
+        ESP_LOGE(LORA_API_TAG, "SPI write failed: reg=0x%02X, val=0x%02X", reg, val);
+        return API_SPI_ERROR;
+    }
+}
+
+api_status_t spi_write_buf(uint8_t reg, uint8_t *val, uint16_t len)
+{
+    uint8_t *out;
+    esp_err_t ret;
+
+    out = (uint8_t *)malloc(len + 1);
```

Comment:
But there is no maximum buffer size to be sent over SPI is there? We can use a value that is the maximum we send but idk if that should be the value

Reply:
There is no inherent buffer limit for SPI. Often, SPI is used to connect external flash memory. However, if you can - and in this case, we do - there should be an assumption of a maximum value according to the device connected to a specific SPI. The main rule is to avoid dynamic memory allocation whenever possible. Sometimes, it's unavoidable, but in such cases, we should create our own allocator that provides a memory pool and prevents memory fragmentation.

Reply:
Right, our own allocator 😅, for now I've just put a hard limit of 128 bytes because that's plenty for Lora use. I've talked with Pawel and there's still some work to do here. To not allocate 128 bytes x2 for an incoming and outgoing buffer. I'll try to fix it when I have some time